xclim/core/indicator.py

```diff
@@ -79,56 +161,184 @@ class Indicator:
+    long_name = ""
+    units = ""
+    missing = "from_context"
+    missing_options = None
+    cell_methods = ""
+    description = ""
+    context = "none"
+    title = ""
+    abstract = ""
+    keywords = ""
+    references = ""
+    comment = ""
+    notes = ""
+
+    def __new__(cls, **kwds):
+        """Create subclass from arguments."""
+        identifier = kwds.get("identifier", getattr(cls, "identifier"))
+        if identifier is None:
+            raise AttributeError("`identifier` has not been set.")
+
+        # Convert function objects to static methods.
+        for key in cls._funcs + cls._cf_names:
+            if key in kwds and callable(kwds[key]):
+                kwds[key] = staticmethod(kwds[key])
+
+        # Parse `compute` docstring to extract missing attributes
+        # Priority: explicit arguments > super class attributes > `compute` docstring info
+
+        func = kwds.get("compute", None) or cls.compute
+        attrs = {
+            k: v for (k, v) in parse_doc(func.__doc__).items() if not getattr(cls, k)
+        }
+        attrs.update(kwds)
+
+        # Create new class object
+        new = type(identifier.upper(), (cls,), attrs)
+
+        # Set the module to the base class' module. Otherwise all indicators will have module `xclim.core.indicator`.
+        new.__module__ = cls.__module__
+
+        #  Add the created class to the registry
+        cls.register(new)
+
+        # This will create an instance from the new class and call __init__.
+        return super().__new__(new)
+
+    @classmethod
+    def register(cls, obj):
+        """Add subclass to registry."""
+        name = obj.__name__
+        if name in registry:
+            warnings.warn(f"Class {name} already exists and will be overwritten.")
+        registry[name] = obj
 
     def __init__(self, **kwds):
+        """Run checks and assign default values.
+        """
 
-        # Set instance attributes.
-        for key, val in kwds.items():
-            setattr(self, key, val)
+        # Check identifier is well formed - no funny characters
+        self.identifier = kwds.pop("identifier", self.identifier)
+        self.check_identifier(self.identifier)
 
-        # Verify that the identifier is a proper slug
-        if not re.match(r"^[-\w]+$", self.identifier):
-            warnings.warn(
-                "The identifier contains non-alphanumeric characters. It could make life "
-                "difficult for downstream software reusing this class.",
-                UserWarning,
+        if self.missing == "from_context" and self.missing_options is not None:
+            raise ValueError(
+                "Cannot set `missing_options` with `missing` method being from context."
             )
 
-        # Default value for `var_name` is the `identifier`.
-        if self.var_name == "":
+        # Validate hard-coded missing options
+        kls = MISSING_METHODS[self.missing]
+        self._missing = kls.execute
```

Comment:
Same comment as in #485: if I understand correctly, this line will break if self.missing is 'from_context', no? It will invoke the empty `FromContext` class.

Reply:
This deserves to be tested, but in principle it should execute the missing algorithm set by options.

Reply:
`from_context` is the default value for the missing attribute. Added a test for this.